File: packages/typescript/src/formatter.ts
import type { Credential } from './types.js';

/**
 * Convert a credential into HTTP headers.
 * Matches signet CLI's strategy.applyToRequest() logic exactly.
 */
export function formatHeaders(credential: Credential): Record<string, string> {
  switch (credential.type) {
    case 'cookie': {
      const headers: Record<string, string> = { ...credential.xHeaders };
      headers['Cookie'] = credential.cookies
        .map(c => `${c.name}=${c.value}`)
        .join('; ');
      return headers;
    }
    case 'bearer': {
      const headers: Record<string, string> = { ...credential.xHeaders };
      headers['Authorization'] = `Bearer ${credential.accessToken}`;
      return headers;
    }
    case 'api-key': {
      const value = credential.headerPrefix
        ? `${credential.headerPrefix} ${credential.key}`
        : credential.key;
      return { [credential.headerName]: value };
    }
    case 'basic': {
      const encoded = Buffer.from(
        `${credential.username}:${credential.password}`
      ).toString('base64');
      return { Authorization: `Basic ${encoded}` };
    }
  }
}

/**
 * Extract localStorage values from a credential.
 * Only cookie and bearer credentials can have localStorage.
 */
export function extractLocalStorage(credential: Credential): Record<string, string> {
  if (credential.type === 'cookie' || credential.type === 'bearer') {
    return { ...(credential.localStorage ?? {}) };
  }
  return {};
}

File: packages/python/src/signet_sdk/formatter.py
import base64
from .types import Credential

def format_headers(credential: Credential) -> dict[str, str]:
    if credential.type == "cookie":
        headers = dict(credential.xHeaders)
        headers["Cookie"] = "; ".join(f"{c.name}={c.value}" for c in credential.cookies)
        return headers
    elif credential.type == "bearer":
        headers = dict(credential.xHeaders)
        headers["Authorization"] = f"Bearer {credential.accessToken}"
        return headers
    elif credential.type == "api-key":
        value = f"{credential.headerPrefix} {credential.key}" if credential.headerPrefix else credential.key
        return {credential.headerName: value}
    elif credential.type == "basic":
        encoded = base64.b64encode(f"{credential.username}:{credential.password}".encode()).decode()
        return {"Authorization": f"Basic {encoded}"}
    return {}

def extract_local_storage(credential: Credential) -> dict[str, str]:
    if credential.type == "cookie" or credential.type == "bearer":
        return dict(credential.localStorage)
    return {}
